src/clients/bridges/BundleDataClient.ts

```diff
@@ -0,0 +1,363 @@
+  // For instance, if the spoke or hub clients have been updated, it probably makes sense to clear this to be safe.
+  clearCache(): void {
+    this.loadDataCache = {};
+  }
+
+  loadDataFromCache(key: string): DataCacheValue {
+    // Always return a deep cloned copy of object stored in cache. Since JS passes by reference instead of value, we
+    // want to minimize the risk that the programmer accidentally mutates data in the cache.
+    return _.cloneDeep(this.loadDataCache[key]);
+  }
+
+  async getPendingRefundsFromValidBundles(bundleLookback: number): Promise<FillsToRefund[]> {
+    const refunds = [];
+    if (!this.clients.hubPoolClient.isUpdated || this.clients.hubPoolClient.latestBlockNumber === undefined) {
+      throw new Error("BundleDataClient::getPendingRefundsFromValidBundles HubPoolClient not updated.");
+    }
+
+    let latestBlock = this.clients.hubPoolClient.latestBlockNumber;
+    for (let i = 0; i < bundleLookback; i++) {
+      const bundle = this.clients.hubPoolClient.getLatestFullyExecutedRootBundle(latestBlock);
+      if (bundle !== undefined) {
+        // Update latest block so next iteration can get the next oldest bundle:
+        latestBlock = bundle.blockNumber;
+        refunds.push(await this.getPendingRefundsFromBundle(bundle));
+      } else {
+        break;
+      } // No more valid bundles in history!
+    }
+    return refunds;
+  }
+
+  // Return refunds from input bundle.
+  async getPendingRefundsFromBundle(bundle: ProposedRootBundle): Promise<FillsToRefund> {
+    // Reconstruct latest bundle block range.
+    const bundleEvaluationBlockRanges = getImpliedBundleBlockRanges(
+      this.clients.hubPoolClient,
+      this.clients.configStoreClient,
+      bundle,
+      this.chainIdListForBundleEvaluationBlockNumbers
+    );
+    const { fillsToRefund } = await this.loadData(bundleEvaluationBlockRanges, this.spokePoolClients, false);
+
+    // The latest proposed bundle's refund leaves might have already been partially or entirely executed.
+    // We have to deduct the executed amounts from the total refund amounts.
+    return this.deductExecutedRefunds(fillsToRefund, bundle);
+  }
+
+  // Return refunds from the next valid bundle. This will contain any refunds that have been sent but are not included
+  // in a valid bundle with all of its leaves executed. This contains refunds from:
+  // - Bundles that passed liveness but have not had all of their pool rebalance leaves executed.
+  // - Bundles that are pending liveness
+  // - Not yet proposed bundles
+  async getNextBundleRefunds(): Promise<FillsToRefund> {
+    const futureBundleEvaluationBlockRanges = getWidestPossibleExpectedBlockRange(
+      this.chainIdListForBundleEvaluationBlockNumbers,
+      this.spokePoolClients,
+      getEndBlockBuffers(this.chainIdListForBundleEvaluationBlockNumbers, this.blockRangeEndBlockBuffer),
+      this.clients,
+      this.clients.hubPoolClient.latestBlockNumber ?? 0,
```

Comment:
Included this to escape the fact that `latestBlockNumber` can potentially be undefined. I feel comfortable adding `?? 0` since we don't do any checks currently.